[TEP-0138]: Decouple API and Feature Versioning Proposal (#1034)
tektoncd/community: teps/0138-decouple-api-and-feature-versioning.md

This TEP proposes to update validations for exisiting beta feature immediately given that this has led to a bug where users are not able to create a v1 TaskRun if they are creating a Task with beta featurs with CRD specified as v1beta1. And after the change, beta features in v1beta1 would also require `enable-api-fields` to be set to `beta` to use.

### Update api compatibility policy
This TEP also proposes to update the [api compatibility policy](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md) to reflect that in order to turn on beta features, it would require the `enable-api-fields` to be set to `beta` regardless of the CRD apiVersion. New features that are promoted to beta stability will also follow this policy, regardless of the CRD's API version. It will clarify that feature stability levels are not correlated with CRD API versions.

### Deprecate the default `beta` stability level

> This TEP proposes that once we are able to decouple api and feature versioning, we should switch back to `stable` as the default `enable-api-fields` value. This would allow us to deprecate the current default value for opt-in features of `beta` stability level.

Reviewer: What's the reason to move back to enable-api-fields = stable by default? This would be another breaking change.

We previously had enable-api-fields = stable and changed it to beta, but this didn't change the fact that beta features were enabled by default for the v1beta1 api. This would make beta features disabled by default for the v1beta1 API. I don't see a strong reason to make this change, and I also don't think this decision needs to happen as part of this TEP.

Reviewer 2: I agree with @JeromeJu that we need to switch back the flag default to "stable" - having the default set to "beta" was a workaround for an issue generated by the decoupling, but it's not something we should keep.
Unfortunately, as you noted, it is a breaking change, but only for API features that are currently in beta. 
What we should do is deprecate the default value being beta, as @JeromeJu suggested, and after 9 months, or as soon as those beta features are stable (whichever happens first) switch the default to "stable".

I think the deprecation itself should be part of this TEP, but the switch to stable does not need to be, we can create an issue for use to remember about it.

Reviewer: @afrittoli can you elaborate on your thinking re: why "stable" is a more appropriate default than "beta"? It sounds like you prefer this approach because the default was previously "stable", but I'd note that for the v1beta1 API, beta features have always been on by default, and this change would turn beta features off by default.

I'm also a bit confused about what it means to deprecate the default value of a flag, rather than to deprecate the flag itself. Since we're not removing the flag and cluster operators can change its value at any time, what's the benefit of announcing a change 9 months in advance? If the deprecation is part of this TEP, doesn't that also imply that this TEP includes a decision to change the default value to stable?

I think the questions of "what's the most appropriate value for enable-api-fields, and is it worth making a breaking change to set enable-api-fields to what we believe is the most appropriate value?" can be tackled separately from this TEP.

Author: Thanks for the discussion on this @afrittoli  and @lbernick both offline and through the comments! 🙏 

 I think first we all agree that with what this TEP proposes, we would be able to default `stable` as the `enable-api-fields` since it was because of the decoupling of api and feature versioning we had to switch to `beta` as the default `enable-api-fields`. But how and when to do so remains to be discussed, and I have also learned the following points:
- We need to decide what is the value for `enable-api-fields` in the long run for pipelines.
- Whether we are going to make this as a backward incompatible change or not remains a question since we need to  gather info from users and cluster operators standpoints on whether giving notice really benefits them.

Therefore, I have opened https://github.com/tektoncd/pipeline/issues/6948. And I am hoping to reach a conclusion there on what's the next step after we decouple the api and feature versioning once TEP0138 is implemented.

---

> status: implemantable

Reviewer: ```suggestion
status: implementable
```

- For maintainers, the maintenance operation of swapping the storage version from v1beta1 to v1 should not have affected our users. However, we had to [change the user-facing default value of enable-api-fields from `stable` to `beta` ](https://github.com/tektoncd/pipeline/pull/6732) before changing the storage version of the API to [avoid breaking PipelineRuns using `beta` features](https://github.com/tektoncd/pipeline/pull/6444#issuecomment-1580926707).

- When promoting features, it could cause confusions for contributors to be dependent on the fact whether an apiVersion is available. For example, during [the promotion to beta for projected workspaces](https://github.com/tektoncd/pipeline/pull/5530), the `v1` api's existence led to confusions of what to do with `beta` features in `v1beta1` and its difference with in `v1`.

## Goals

> - Feature validations and implementation should be independant from any API version.

Reviewer: ```suggestion
- Feature validations and implementation should be independent from any API version.
```

> - Come up with a backward-compatible migration plan for setting the enable-api-fields feature flag to stable in the long term.

Reviewer: The proposed solution changes enable-api-fields back to stable, but not in a backwards compatible way. Can this goal be reworded?

Author: I think I understand the point of switching back from "beta" to "stable" `enable-api-fields` is backward incompatible. But I might be wrong understanding the backwards incompatibility. Can I check if we provide with users the option to switch to another value so that they can continue to use the previous behaviour, can we say this solution is backward compatible?

(also related with the comment below)

Reviewer: typically we use "backwards compatible" to mean that an api call that succeeded before the change will still succeed after the change (more details on how this applies in k8s here https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility). In this case, if you want to create a taskrun using beta features, it will succeed before the change but not after the change, unless the cluster operator modifies defaults. so I'd say this is not a backwards compatible change, but we can provide options to make the migration easier.

Author: Thanks for the clarifications! I updated it to providing an easier migration plan instead of it being backward compatible.

- For maintainers, the maintenance operation of swapping the storage version from v1beta1 to v1 should not have affected our users. However, we had to [change the user-facing default value of enable-api-fields from `stable` to `beta` ](https://github.com/tektoncd/pipeline/pull/6732) before changing the storage version of the API to [avoid breaking PipelineRuns using `beta` features](https://github.com/tektoncd/pipeline/pull/6444#issuecomment-1580926707).

- When promoting features, it could cause confusions for contributors to be dependent on the fact whether an apiVersion is available. For example, during [the promotion to beta for projected workspaces](https://github.com/tektoncd/pipeline/pull/5530), the `v1` api's existence led to confusions of what to do with `beta` features in `v1beta1` and its difference with in `v1`.

## Goals

- Feature validations and implementation should be independent from any API version.
- Come up with a plan that makes the migration easier for setting feature flags to enable stable features only by default in the long term.

> - Changes and updates made to the existing feature validaitons regarding decoupling API and feature versioning should keep much backwards compatiblity as possible.

Reviewer 3: `s/validaitons/validations`

For example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: feature-flags
data:
  # alpha: v0.53
  #
  # Pipeline in pipeline has ... functionalities.
  # It is disabled by default.
  enable-pipeline-in-pipeline: "false"
  # alpha: v0.53
  # 
  # Trusted artifacts has ... functionalities.
  # It is disabled by default. Set it to true to enable this feature.
  enable-trusted-artifacts: "true"
```

See [implementation plan](#implementation-plan) for more details on the PerFeatureFlag struct.

All **new** features can only be enabled via per-feature flags. When they first get introduced as `alpha`, they will be disabled by default. When new features get promoted to `stable`, they will be enabled by default according to the following table:

| Feature stability level | Default  |
| ----------------------- | -------- |
| Stable                  | Enabled; Cannot be disabled  |
| Beta                    | Disabled |
| Alpha                   | Disabled |

> Note that per-feature flags that have stabilized cannot be disabled. We will deprecate the per-feature flag after it has become stable and then remove it eventually after the deprecation period according to the API compatibility policy. We will give deprecation and later removel notice of the per-feature flags via release notes after their promotion to `stable`. Cluster operators who do not want such opt-in features would have enough notice to implement admission controllers on their own to disable the feature.

Reviewer 2: @JeromeJu sorry, I thought I commented on this before, but I must have missed it 🙏 

Keeping the feature flag around, deprecated but unusable does not make much sense to me.
Making it unusable provides the same behaviour as removing it, the deprecation has no effect really.

The only reason I could see to keep the flag around would be if we started validating the content of the config file (which we do not do today) - but even so, I think failing config validation would be preferrable to silently ignoring a feature flag.

Reviewer 2: Never mind, I think I misread this section 🙏
A small update to the table might help  - but feel free to ignore this - I can also propose a PR with the update if that's ok for you.

```
| Feature stability level | Default  |
| ----------------------- | -------- |
| Stable                  | Enabled; Cannot be disabled once the flag is removed (after deprecation) |
| Beta                    | Disabled |
| Alpha                   | Disabled |
```

Author: Thanks for the suggestion, PTAL https://github.com/tektoncd/community/pull/1091 🙏

- A new option, `enable-api-fields` = `new-stable`, will be added to the API. This option will use the preferred validation where `stable` and `beta` features are validated the same across apiVersions. In 9 months, `new-stable` will be renamed as `stable`.
- The current behavior for “stable” `enable-api-fields` remains the same for now where it allows users to use `beta` features in `v1beta1` that have coupled feature and api versioning in `v1beta1` i.e. remote resolution. In 9 months “stable” will be renamed as “legacy-stable” and be marked as deprecated.
- The existing `beta` option will remain the same for now and in 9 months.

- Taking the remote resolution feature, which currently couples feature and api versioning, as an example, after the change:
  - With `new-stable` in `v1beta1`, we disable resolver.
  - With `legacy-stable` in `v1beta1`, we do not validate resolvers as a field needs `enable-api-fields` as beta, so they are still turned on by default.
  - When `enable-api-fields` is set to `beta` in both `v1` and `v1beta1`, we are turning on resolver.

### Provide an immediate solution to decouple api and feature versioning now

> This solution proposes that we provide an immediate solution to turn on stable features only with `stable` `enable-api-fields` and require `enable-api-fields` to be set to `beta` when using beta features, regardless of CRD API version. This will solve the unintended behavior that taskruns cannot be created for version coupled features with “enable-api-fields” set to “stable” during the [v1 storage swap](https://github.com/tektoncd/pipeline/pull/6444#issuecomment-1580926707) right away.

Reviewer: Can you reword this sentence a little bit? This sentence makes it sound like the behavior of "enable-api-fields=stable" is changing immediately, but the "stable" value isn't changing for 9 months. What's changing now is that we're providing a new value to address the problem.

Author: Thanks, that makes sense! Updated.

### Migrate `enable-api-fields` to `stable` in 9 months
The default value for `enable-api-fields` will be switched back to `stable` in the long run, as discussed in [issue #6948](https://github.com/tektoncd/pipeline/issues/6948). The `new-stable` flag introduced will be renamed to `stable` in 9 months, at which point it will become the default value. And the current `stable` behavior will be renamed to `legacy-stable` for backwards compatibility. The `legacy-stable` flag will be deprecated and then removed when `v1beta1` is removed or once all current `beta` features that are coupled(object params and results, array results and indexing and remote resolution) become `stable`, whichever happens first.

## Design Evalutaion
### Pros

> - The proposed solution provides immediate options for users who want to use only stable features in `v1beta1`. This helps to prevent adding more complexity to the existing validation if more `beta` features are being added to `v1beta1`.

Reviewer: I think this new option is actually more beneficial to cluster operators who want to restrict usage to only stable features, regardless of what apiversion their users have defined in their crds

> - Workarounds are provided instead of having backwards incompatible changes during migration of the defaulting `enable-api-fields` back to `stable`. For example, users that accidentally have `beta` features opt-in with `v1beta1` CRDs could migrate to the usage with `legacy-stable` validation.

Reviewer: Can you reword this? This change is backwards incompatible, we're just providing more warning and the option to retain legacy behavior.

Author: Updated! Thanks again for the clarifications on the backwards compatibility.

### Cons

> - The introduced new flag will add complexity to the testing matrix. We will also need to introduce tests for `new-stable`, and to keep testing `beta` and `legacy-stable` in 9 months.

Reviewer 3: I wouldn't focus on the "complexity" added to the testing matrix, this is not really something a "user" care or have to care, right ?

What the user care is, the user experience. What we are proposing here, is to make people, most likely change twice the value of a flag in one and half year.

### Make `beta` feature validation changes now; migrate `enable-api-fields` to `stable` in 9 months
Require `enable-api-fields` to be set to `beta` when using beta features, regardless of CRD API version. Immediately make this change for existing beta features. This will solve the unintended behavior that taskruns cannot be created for version coupled features with “enable-api-fields” set to `stable` during the `v1` storage swap right away.

For the default value of `enable-api-fields` in the long run, this alternative will give a nine month notice for users for the breaking change setting the default to `stable`.

#### Pros
- This solution could provide best clarity of feature stability levels across different apiVersions by keeping validation for beta features the same in `v1` and `v1beta1`.

#### Cons
- Updating existing beta features valiations in `v1beta1` is a backwards incompatible change. This will break users who are accidentally using the coupled beta features while `enable-api-fields` is set to stable.

### New `legacy-stable` value for `enable-api-fields`; migrate `enable-api-fields` to `stable` in 9 months

> Change the validation for existing default `beta` value of `enable-api-fields` to validate all beta features across apiVersions. Add a `legacy-stable` value to the `enable-api-fields` to keep the behavior of the current "stable" feature validations in `v1beta1` that includes the coupled features. 

Reviewer: I don't think this solution would involve any changes when enable-api-fields is set to "beta"; instead, it would change the validation of beta features when enable-api-fields is set to "stable".

Author: That makes sense, I changed the wording regarding stable.

This alternative proposes that we will keep the current implementations and validations for beta features in `v1` and `v1beta1` apiVersion and only make new features have the synchronized validations across different apiVersions.

#### Pros
- This is backwards compatible.
- Changes are minimal for the current codebase with probably only documentation required.
#### Cons
- We would still need to use `beta` as the default stability of features that are turned on. Some beta features are still coupled in feature and api versioning. This will not allow cluster operators to opt-in only stable features in v1beta1.

### Give 9-month warning before making v1beta1 validation changes and default to stable
This alternative proposes the validation change for beta features in v1beta1 to be done with giving a notice for the breaking change for 9 months.

#### Cons:
- Since v1beta1 only has 12 months of support period left, making a breaking change to it might have more impacts on v1beta1 users than its worth after 9 months.

### Wait until v1beta1 is removed to swap `enable-api-fields` back to `stable`
#### Pros
- Straightforward and no code changes need to be done.

#### Cons

> - This is still a breaking change for v1 even when v1beta1 deprecates.

Reviewer: ```suggestion
- This is still a breaking change for v1 even when v1beta1 is removed.
```

- It should have a testing stratey that will give us confidence in our impelmentations of per feature flags and changes to existing feature flags.

## Proposal

This TEP provides a plan for ensuring that feature stability doesn't depend on CRD API version, and using per-feature flags to move to a future where only stable features are enabled by default through the following steps:
- [Change existing validation to decouple feature and API versioning](#change-existing-validation-to-decouple-feature-and-api-versioning)
- [Per feature flag for new api-driven features](#per-feature-flag-for-new-api-driven-features)
- [Sunset `enable-api-fields` after existing features stabilize](#sunset-enable-api-fields-after-existing-features-stabilize)

## Design Details

### Change existing validation to decouple feature and API versioning
We will change the current validation for `enable-api-fields=stable` to only allow using stable features regardless of API version. This will resolve the current issue of the coupling of API and feature versioning in v1beta1. More specifically, beta features resolvers, object/ array params and results will require `enable-api-fields` set to `beta` to be used.

> Note that although this is a behavior change, it is more of a bug fix for the coupling of feature and API versioning. Currently, with `enable-api-fields` set to `stable`,  PipelineRuns like [this one](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/git-resolver.yaml) fail because the controller cannot create child TaskRuns. This change will result in a validation failure instead.

Reviewer 3: Can we add something to give an understanding on what it means ? Something like :

This would mean that, today, to use resolvers, you need to set `enable-api-fields` to `false`, even if you are using `v1beta1` APIs (for `PipelineRun`, …). The default value being `beta`, it wouldn't break users *by default*, but it would change a behavior for those who only wanted to enable `stable` features.

Author: Added. Also we have included the impacts on users below for elaborating those use cases. Hope that helps)

For example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: feature-flags
data:
  # alpha: v0.53
  #
  # Pipeline in pipeline has ... functionalities.
  # It is disabled by default.
  enable-pipeline-in-pipeline: "false"
  # alpha: v0.53
  # 
  # Trusted artifacts has ... functionalities.
  # It is disabled by default.
  enable-trusted-artifacts: "true"
```

See [implementation plan](#implementation-plan) for more details on the PerFeatureFlag struct.

All **new** features can only be enabled via per-feature flags. When they first get introduced as alpha, they will be disabled by default. When new features get promoted to stable, they will be enabled by default according to the following table:

| Feature stability level | Default  |
| ----------------------- | -------- |
| Stable                  | Enabled; Cannot be disabled  |
| Beta                    | Disabled |
| Alpha                   | Disabled |

> Note that for per feature flag that has stabilized, they cannot be disabled and we will remove the feature flag after 3 releases after it has become stable. Notification will be given to users via release notes.

Reviewer 3: Not sure if it's covered by our API rules of 9 months, but by releases here, do we mean LTS or any releases ? and if any, how do we take LTS into account ? 

In addition, the flag "might" be removed (aka being there or not shouldn't have an effect, nor its value) but I wonder if we should "validate" thoses ? (aka fail if a feature-flag is no more)

Author: I agree that the removal of the flag should have no effect on the 'stable' feature usage since they cannot be disabled in any way. 

> but I wonder if we should "validate" thoses
I would like to clarify, by "validate" you mean validate features with their per feature flag at "stable" level or do you mean we should validate them after the flags are removed? If it is the first one, implementation-wise, we have added a fields, "LockToDefault" (similar to what k8s does now), indicating that we might not even need to "validate" stable features since they are locked to be turned on by default.

> Not sure if it's covered by our API rules of 9 months, but by releases here, do we mean LTS or any releases ?
That's a great point. I intended to mean 3 'general' releases (either LTS or non-LTS). The idea here is any 3 releases shall include 1 LTS, so at least one feature remains stable in 1 LTS the feature flag could be removed. 

In terms of API compatibility policies, I think this is exactly the new part and we would need to reach agreement on. I would lean towards that this doesn't need to fall under the 9 months' deprecation rule since it does not affect users in any case. Happy to hear from your opinions on this.

Reviewer 4: Let's say a feature is marked as stable in `0.53`, the flag can potentially be deleted in `0.57` after three releases (`0.54` LTS, `0.55`, and `0.56`). Now, the next LTS is `0.58`. If a cluster is upgraded from `0.54` to `0.58`, do we expect the cluster operator to delete that feature flag from their deployment script?

Reviewer 2: If an operator enables a feature `foobar` before it's stable, they will have the `enable-foobar` flag set to true. Once the feature is promoted to stable, they should be able to continue using the feature seamlessly, i.e. we should ignore `enable-foobar=true`.
 
If an operator disables a feature `foobar` before it's stable, once the feature becomes stable they will have to decide what to do. If they still don't want users to use that feature they may employ a custom admission controller to do so. I believe the best user experience would be to be notified that the flag is deprecated and will be removed in X releases so that they have time to develop an alternate solution - that requires the flag to continue work for some time after the flag reaches `stable`. 

If an operator does not set the `enable-foobar` flag (they use the default value), the switch to `stable` will be transparent to them. They will be aware through release notes that a new API feature is available.

I don't think we should keep the flag around if it cannot be used. I would recommend deprecating it instead and then removing it after the deprecation period.

> do we expect the cluster operator to delete that feature flag from their deployment script

@pritidesai I think we should issue a warning if the flag is set to `false` (i.e. feature disabled) starting with `0.53`. From `0.58` we can apply the behaviour we apply for unknown flags (I'm not sure if we fail or ignore?)

Author: I agree with the suggestion that we are deprecating stabilized feature flags and removing it eventually. This shall fully comply with our API policies and giving users enough notice from release notes to react.

Note that although this is a behavior change, it is more of a bug fix for the coupling of feature and API versioning. Currently, with `enable-api-fields` set to `stable`,  PipelineRuns like [this one](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/git-resolver.yaml) fail because the controller cannot create child TaskRuns. This change will result in a validation failure instead.

However, the default value of `enable-api-fields` will **not** be changed and it will remain to be `beta`, so the **beta** features will continue to be enabled by default.

- **Impacts on users:**
    - Cluster operators:
      - Current cluster operators with `enable-api-fields` set to `alpha` or `beta` should not experience any changes.
      - This makes it possible for cluster operators to have full control over only stable feature usages, rather than user overrides. For example, currently if cluster operators want their users to only opt-in “stable” features, they cannot do so for v1beta1 apiVersion for the exception of resolver, object params and results.
    - Pipeline and Task authors: 

>     This affects pipeline users in the situation where cluster operators have chosen to set `enable-api-fields` to `stable` (i.e. those who have changed the current default `beta` value) and who are accidentally using beta features, like resolvers, in v1beta1.

Reviewer 4: I am generally not in favor of using this term (`accidentally`) here, specially in the TEP proposal.

Let's try to explain what this accident mean here.

`v1beat1` Pipeline definitions using `resolver` is applied to a cluster with `enable-api-fields` set to `stable`, is it? But this `pipeline` does not result in a successful `pipelineRun` because of <some reason>.  The resolver is still a `beta` feature and this pipeline should not have been applied to the cluster in first place. Does this make sense?

Author: Thanks Priti! I've added the examples and reasoning for elaborating this.

### Per feature flag for _new_ api-driven features
Introduce per-feature flags for each **new** API driven feature. Each feature will have its own flag, instead of using the group API driven flag `enable-api-fields` to enable or disable all features of a stability level. Note that our proposal only takes effect on api-driven features while behavioural flags will remain the existing beahviour.

The flag will also include the maintainer information and stability level for the feature as the source of truth. New behavioural features will also have a new per-feature flag to either enable or disable the feature. This will allow behavioural features that has values leading to behaviors in different stability levels to be turned on or off instead of depending on the stability level of features that are enabled.

For example:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: feature-flags
data:
  # alpha: v0.53
  #
  # Pipeline in pipeline has ... functionalities.
  # It is disabled by default.
  enable-pipeline-in-pipeline: "false"
  # alpha: v0.53
  # 
  # Trusted artifacts has ... functionalities.

>   # It is disabled by default.

Reviewer 4: I think this meant `enabled` and not `disabled`, is it?

> ### Per feature flag for _new_ api-driven features

Reviewer 5: It seems like the API versions and feature stability levels are already decoupled with the [Change existing validation to decouple feature and API versioning](https://github.com/tektoncd/community/blob/a5462c5dc565883646f8c6b719c07f046fa34999/teps/0138-decouple-api-and-feature-versioning.md#change-existing-validation-to-decouple-feature-and-api-versioning) proposal. I'm wondering what is the motivation to use per-feature flag model?

Author: Thanks for the comment. I've put back the missing pieces:)

The reason that this was not previously included was that we first aimed to decouple feature versioning only and while moving forward with the discussion in API WG we've decided to use per-feature flag as the solution to decouple versioning in order to avoid the potential issues from setting the default stability level in the long run.

Reviewer 5: Thanks!

> Introduce per-feature flags for each **new** API driven feature. Each feature will have its own flag, instead of using the group API driven flag `enable-api-fields` to enable or disable all features of a stability level. Note that our proposal only takes effect on api-driven features while behavioural flags will remain the existing behaviour.

Author: Thanks for the suggestion. Added.

  - We cannot enable existing features using per-feature flags because this would not be backwards compatible. If we allowed this, the individual flag would have precedence over the grouped flag, which means that to preserve backwards compatibility, the individual flag would need to be on by default for beta features and off by default for alpha features. However, this would not be backwards compatible for cluster operators who set `enable-api-fields` to `stable`, since they would also need to override the new `beta` level per feature flags. 

- **Cluster Operators:** For new features, cluster operators will explicitly turn on or off each features in the ConfigMap. They will be able to choose to turn on a single feature. See future work for how cluster operators would communicate with their users on the list of features enabled.

- **Task and Pipeline Authors:** Tekton Pipeline and Task authors will get to know the list of features that are turned on from their service providers. See [future work](#future-work) for the better communication from cluster operators to users for more details.

> ### Sunset `enable-api-fields` after existing features stabilize

Author: I agree that we shall give heads up for existing features. But I am not sure if keeping it in the current TEP would be helpful since basically all new features coming in after this TEP being merged shall use per-feature flag model ideally.(Plus merging this TEP sooner so to avoid such situations) Maybe giving heads up in the WG could be a nice option WDYT?

> ### Sunset `enable-api-fields` after existing features stabilize

Reviewer 5: Do you mean that new features coming in should follow per-feature flag after this TEP is `merged` or `implemented`?

> ### Sunset `enable-api-fields` after existing features stabilize

Author: Yes:) And it looks like most maintainers are on board with this TEP, we are hoping to get this merged by EOD. I will go ahead and update the API-compatibility policy ASAP.

This alternative proposes the validation change for stable features in v1beta1 to be done with giving a notice for the breaking change for 9 months.

#### Cons:
- Since v1beta1 only has 12 months of support period left, making a breaking change to it might have more impacts on v1beta1 users than its worth after 9 months.

## Implementation Plan

For each new feature driven by api fields, a new feature flag using FeatureFlag struct will be added to the existent set of feature flags. This struct will include the stability of feature and whether it is turned on or locked to default. Currently it will only work for new features without `enable-api-fields`, see future work for more details of the option to introduce `enable-api-fields=none` to use per feature flags for new features.

```go
type FeatureFlag struct {
    // Name of the feature flag
    Name string
    
    // Stability level of the feature, one of "stable", "beta" or "alpha"
    Stability string
    
    // Enabled is whether the feature is turned on
    Enabled bool
    
    // LockToDefault indicates whether the feature is locked to default value and cannot be changed

>     LockToDefault bool

Reviewer 6: why do we need this?

Author: This is intended to indicate that the feature flags for the features that have become stable after its graduation process could be removed. We are hoping to have this field as a sign telling its users that it is enabled by default.
Hope this could be helpful; happy to hear suggestions

Reviewer 6: ok, sounds fair but I wasn't sure if it was absolutely necessary!

Reviewer 2: I think the flag should either be there or not, I don't see the value of having the flag locked so that in fact it cannot be used.

Reviewer 2: I think the flag should either be there or not, I don't see the value of having the flag locked so that in fact it cannot be used.

Author: After the rounds of feedback in this TEP and considerations, I agreed to remove this field. And with the suggestions here https://github.com/tektoncd/community/pull/1034#discussion_r1317175511, when we are deprecating a flag after it stabilized and removing it eventually.

    
    // Deprecated is whether the feature is deprecated
    // +optional
    Deprecated bool
}
```

## Test Plan
### Retain existing CI end-to-end testing matrix 
We plan to retain the existing testing matrix for fields at `alpha`, `beta` and `stable`. With the addition of per-feature flags for new features, it would look like:

| |enable-api-fields|Per feature flags|Integration tests|
| -- | -- | -- | -- |
|Opt-in stable| stable|Turn OFF all per feature flags|Run all stable e2e tests.|
|Opt-in beta| beta|Turn ON all beta per-feature flags|Run all beta e2e tests.|
|Opt-in alpha| alpha|Turn ON all per-feature flags (alpha and beta)|Run all alpha e2e tests.|

## Additional CI tests

> ### Additional Test Combinations

Reviewer 6: is this in addition to testing with all alpha/beta/stable features on?

Author: Yes, these are for the purpose of testing the combinations of per feature flags.

Reviewer 4: I am having hard time understanding the following table. How do we choose a small number of `e2e` tests? What are we achieving with this kind of testing?

We might have to think this a little more, its not a blocker but something we need to address as part of this proposal.

How is Kubernetes (an inspiration for this proposal) handling it?

Some key points for us to think about:

The cluster operator can enable one or more features using per feature flags. Pipeline/task authors can choose to leverage none or only one or multiple of the enabled features in their specifications. As a project maintainer, we have to make sure a feature works in isolation and also does not impact any existing functionalities.

This TEP provides a plan for ensuring that feature stability doesn't depend on CRD API version, and using per-feature flags to move to a future where only stable features are enabled by default through the following steps:
- [Change existing validation to decouple feature and API versioning](#change-existing-validation-to-decouple-feature-and-api-versioning)
- [Per feature flag for new api-driven features](#per-feature-flag-for-new-api-driven-features)
- [Sunset `enable-api-fields` after existing features stabilize](#sunset-enable-api-fields-after-existing-features-stabilize)

## Design Details

### Change existing validation to decouple feature and API versioning
We will change the current validation for `enable-api-fields=stable` to only allow using stable features regardless of API version. This will resolve the current issue of the coupling of API and feature versioning in v1beta1. More specifically, beta features resolvers, object/ array params and results will require `enable-api-fields` set to `beta` to be used. This means that, for those current beta features (resolvers, object/ array params and results) will no longer be enabled with `enable-api-fields=stable`. 

> To continue using these features, users will need to explicitly set the `enable-api-fields` flag to `beta`. This change will not affect users who are already using the `enable-api-fields=beta` flag, which is the default and will continue to be. It would change the behaviour for those who only want to enable `stable` features. 

Reviewer 4: ```suggestion
To continue using these beta features, users will need to explicitly set the `enable-api-fields` flag to `beta`. This change will not affect the pipeline deployments with `enable-api-fields` flag set to `beta. This is the default configuration and will continue to be. This will impact the cluster operators who would like to enable only `stable` features. The cluster operators will have to change the configuration of the pipeline deployment such that `enable-api-fields` is set to `stable` for their deployments. 
```

Author: Thanks for adding to the details of the impacts 🙏  updated:)

> Note that although this is a behavior change, it is more of a bug fix for the coupling of feature and API versioning. Currently, with `enable-api-fields` set to `stable`,  PipelineRuns like [this one](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/git-resolver.yaml) fail because the controller cannot create child TaskRuns. This change will result in a validation failure instead.

Reviewer 4: ```suggestion
Note that although this looks like a behavior change, it is actually a bug fix. Currently, with `enable-api-fields` set to `stable`,  PipelineRuns like [this one](https://github.com/tektoncd/pipeline/blob/main/examples/v1/pipelineruns/beta/git-resolver.yaml) fail because the controller cannot create child TaskRuns. This change will result in a validation failure instead and this `pipelineSpec` will be prohibited with ``enable-api-fields=stable`.
```

However, the default value of `enable-api-fields` will **not** be changed and it will remain to be `beta`, so the **beta** features will continue to be enabled by default.

- **Impacts on users:**

>     - Cluster operators:

Reviewer 4: ```suggestion
    - Cluster Operators:
```

      - Current cluster operators with `enable-api-fields` set to `alpha` or `beta` should not experience any changes.

>       - This makes it possible for cluster operators to have full control over only stable feature usages, rather than user overrides. For example, currently if cluster operators want their users to only opt-in “stable” features, they cannot do so for `v1beta1` apiVersion for the exception of resolver, object params and results.

Reviewer 4: ```suggestion
      - This makes it possible for the cluster operators to have a full control over their deployments. The cluster operator can guarantee enabling only `stable` features and avoid user bypassing this enforcement . For example, currently it is not possible for the cluster operators to enable “stable” features for `v1beta1` with an exception of a list of beta features such as resolver, object params and results.
```

Author: Thank you Priti!

>     - Pipeline and Task authors: 

Reviewer 4: ```suggestion
    - Pipeline and Task Authors: 
```

>     This affects pipeline users in the situation where cluster operators have chosen to set `enable-api-fields` to `stable` (i.e. those who have changed the current default `beta` value). 

Reviewer 4: ```suggestion
    - The same `pipeline` and `task` definitions might not be applied to a cluster with `enable-api-fields` set to `stable` after this change is implemented.
    - The `pipeline` and `task` definitions in `v1beta1` implementing [beta features](https://github.com/tektoncd/pipeline/blob/main/docs/additional-configs.md#beta-features) will not be applied to a cluster with `enable-api-fields` set to `stable` Such pipeline .....
```

### Per feature flag for _new_ api-driven features
Introduce per-feature flags for each **new** API driven feature. Each feature will have its own flag, instead of using the group API driven flag `enable-api-fields` to enable or disable all features of a stability level. Note that our proposal only takes effect on api-driven features while behavioural flags will remain the existing beahviour.

The flag will also include the maintainer information and stability level for the feature as the source of truth. New behavioural features will also have a new per-feature flag to either enable or disable the feature. This will allow behavioural features that has values leading to behaviors in different stability levels to be turned on or off instead of depending on the stability level of features that are enabled.

For example:

```yaml
apiVersion: `v1`
kind: ConfigMap
metadata:
  name: feature-flags
data:
  # alpha: v0.53
  #
  # Pipeline in pipeline has ... functionalities.
  # It is disabled by default.
  enable-pipeline-in-pipeline: "false"
  # alpha: v0.53
  # 
  # Trusted artifacts has ... functionalities.
  # It is disabled by default.

>   enable-trusted-artifacts: "true"

Reviewer 4: suggested additional comment: set it to true to enable this feature

Author: Thanks:) updated.

```

See [implementation plan](#implementation-plan) for more details on the PerFeatureFlag struct.

All **new** features can only be enabled via per-feature flags. When they first get introduced as alpha, they will be disabled by default. When new features get promoted to stable, they will be enabled by default according to the following table:

| Feature stability level | Default  |
| ----------------------- | -------- |
| Stable                  | Enabled; Cannot be disabled  |
| Beta                    | Disabled |
| Alpha                   | Disabled |

Note that for per feature flag that has stabilized, they cannot be disabled and we will remove the feature flag after 3 releases after it has become stable. Notification will be given to users via release notes.

The behaviour of existing `enable-api-fields` flag with per feature flag:
- Any current beta features can be enabled with `enable-api-fields` set to “beta” or “alpha”.
- Any current alpha features can be enabled with `enable-api-fields` set to “alpha”. When current alpha features are promoted to beta, they can be enabled with `enable-api-fields` set to `beta` or `alpha`.

> - It will not be possible to enable existing features using per-feature flags.

Reviewer 4: ```suggestion
- The existing features are not adopting this new mechanism of having one flag per feature. It will not be possible to enable existing features using per-feature flags.
```

Author: Thank you for the suggestion!

#### Cons
- We would still need to use `beta` as the default stability of features that are turned on. Some beta features are still coupled in feature and API versioning. This will not allow cluster operators to opt-in only stable features in `v1beta1`.

### Give 9-month warning before breaking changes and default to stable
This alternative proposes the validation change for stable features in `v1beta1` to be done with giving a notice for the breaking change for 9 months.

#### Cons:
- Since `v1beta1` only has 12 months of support period left, making a breaking change to it might have more impacts on `v1beta1` users than its worth after 9 months.

## Implementation Plan

For each new feature driven by api fields, a new feature flag using FeatureFlag struct will be added to the existent set of feature flags. This struct will include the stability of feature and whether it is turned on or locked to default. Currently it will only work for new features without `enable-api-fields`, see future work for more details of the option to introduce `enable-api-fields=none` to use per feature flags for new features.

```go

> type FeatureFlag struct {

Reviewer 2: I guess this will be kept separated from the behavioural feature flags configuration? 
Some of those today support multiple values, which would not fit here.

Author: Yes! All the design parts in this TEP shall be related with features enabled with API fields. Behavioural flags shall already been of per-feature based, so it should not have the implicit possibility to be coupled.

    // Name of the feature flag

>     Name string

Reviewer 2: What is the name used for? Perhaps for logging purposes? 
Perhaps we could do something similar to what we have for behavioural flags:

```
type APIFeatureFlags struct {
   Foo  FeatureFlag
   Bar   FeatureFlag
}
```

Reviewer 2: ?

Author: I agree the name field could be kept for logging purposes.
As for the grouping of all the API feature flags, my understanding is that such struct would make sense if we would like to turn on behavioural features of a certain stability level. But I am not sure if this is going to be necessary while we fully switch to per-feature flags (i.e. after all current beta/alpha features stabilized)?

    
    // Stability level of the feature, one of "stable", "beta" or "alpha"

>     Stability string

Reviewer 2: This would basically only be for documentation purposes since the decision is only based on the `Enabled` part, so I'm not sure we need to have it part of the struct?

Author: This struct attempted to follow what [kubernetes has for their features](https://github.com/kubernetes/kubernetes/blob/56cc5e77a10ba156694309d9b6159d4cd42598e1/staging/src/k8s.io/component-base/featuregate/feature_gate.go#L73). And for the purposes of this field, I think there are two main benefits to have them:
- Previously we haven't had any real "source of truth" to tell what stability of a feature is apart from [additional-configs.md](https://github.com/tektoncd/pipeline/blob/release-v0.51.x/docs/additional-configs.md#alpha-features) which relies on manual edition and they can be moved to elsewhere sometimes (i recall once we had beta feature list in `install.md` , take [v0.41.x](https://github.com/tektoncd/pipeline/blob/release-v0.41.x/docs/install.md) for example). So with this field added, this should be regarded as the only source of truth for telling which stability level a feature is at currently.
- As pointed here, for easier documentation purposes. Maintainers/ Cluster operators would have a better sense of which stability level the specific feature is at.

Reviewer 2: I think it would be interesting to see how that is used by k8s, but I guess it's ok to keep it even if we don't have a use for it yet.

    
    // Enabled is whether the feature is turned on
    Enabled bool
    
    // LockToDefault indicates whether the feature is locked to default value and cannot be changed
    LockToDefault bool
    
    // Deprecated is whether the feature is deprecated
    // +optional
    Deprecated bool
}
```

## Test Plan
### Retain existing CI end-to-end testing matrix 
We plan to retain the existing testing matrix for fields at `alpha`, `beta` and `stable`. With the addition of per-feature flags for new features, it would look like:

| |enable-api-fields|Per feature flags|Integration tests|
| -- | -- | -- | -- |
|Opt-in stable| stable|Turn OFF all per feature flags|Run all stable e2e tests.|
|Opt-in beta| beta|Turn ON all beta per-feature flags|Run all beta e2e tests.|

> |Opt-in alpha| alpha|Turn ON all per-feature flags (alpha and beta)|Run all alpha e2e tests.|

Reviewer 4: is this testing both alpha and beta? or just `alpha`?

Author: - For beta features, we will turn on all beta features and test beta features only.
- For alpha features, we will turn on all features (beta+alpha) 
These are for keeping the existing test suites for `pull-tekton-pipeline-beta-integration-test` and `pull-tekton-pipeline-alpha-integration-test` .

|Opt-in stable|Set to stable|<p>All feature flags are OFF by default.</p><p>Turn ON one feature flag at a time. </p>|Run a small number of e2e tests against N combinations. It is not feasible to run the entire e2e test suite against N combinations.|
|Opt-in beta|Set to beta|Turn ON all beta per-feature flags by default.<br><br>Turn ON one per-feature (flag at an alpha stability level) at a time.|Run a small number of e2e tests against M combinations (where M <= N)|
|Opt-in alpha|Set to alpha|Turn ON all per-feature flags by default.<br><br>TURN OFF one feature flag at a time.|Run a small number of e2e tests against N combinations.|

### <a name="_tcstxie74non"></a>How many tests can we run in a reasonable amount of time?
Based on a [recent PR](https://github.com/tektoncd/pipeline/pull/7032), our integration tests take between 26 mins (stable) → 33 mins (alpha). We don’t want to go beyond that. Based on [Feature flags testing matrix](https://docs.google.com/document/d/1r_MX9-mzRtdbfNQq5VC4guHb-tphA0WxWhlQIsusEEA/edit?resourcekey=0-RALry7-GaKn9i19UEaRnYg) benchmarking, the approximate time is:

T = N<sub>pipelines</sub>\*N<sub>tasks</sub>\*N<sub>features</sub>\*6 s

Assume: 

> N<sub>features</sub> = 20 (currently we have 17 api features; lets assume that at any point in time, we will likely have ~20 individual features)

Reviewer 4: This sounds a little disappointing, we need to do a better job of promoting or deprecating features. But it is reasonable given the maturity of the project.

N<sub>tasks</sub> = 2 (two tasks per pipeline)

|Scenario|N<sub>pipelines</sub>|N<sub>features</sub>|N<sub>tasks/pipeline</sub>|T (mins)|
| :- | :- | :- | :- | :- |
|How many pipelines can we afford to run in 30 mins?|**7**|20|2|30|
|How long would it take to run a single (same) pipeline for all the features?|1|20|2|**4**|

### How many tests should we run against the additional tests?
- It is not feasible to run the entire e2e test suite against N features.
- Coming up with a single test that covers all common core features might be challenging.
  - We probably want our tests to be able to cover
    - params
    - workspaces
    - results
    - DAG (small fan out)?
    - Finally

>     - remote resolution

Reviewer 4: The list feels long but we should follow the practice of testing a real world use cases which involves mostly all of the features in a single pipeline.

- During implementation, we can figure out the number of pipelines that allow us to cover as many areas of the API spec as possible.

For testing out individual per feature flags, we will use unit tests for each single feature flag in end to end tests and the combinations of features that could have overlapped(needs to elaborate). The integration tests for each flag will be in place for the CI while the combinations will be in the nightly tests.

- Testing of behavioural features remains the way they run as of today.

## Future Work
- For seeking better way of communicating enabled features from cluster operators and downstream users including service providers and pipeline authors. CLI feature request to query all enabled features at each stability level via tkn.
    - Example output:
      ```
      $pipeline: tkn features list enabled
      stable: <feature-0>
      beta: <feature-1>, <feature-2>
      alpha: <feature-3>
      ```
- For preserving the ability of easily enabling all features of a stability level, we could provide some script that turns on all alpha/beta features by modifying the group of per feature flags.

> - In order to avoid feature To have the process of documenting the stability of features asn an auto genenrated process according to the Per Feature Flags struct being updated as a source of truth.

Reviewer 4: ```suggestion
- In order to avoid feature to have the process of documenting the stability of features an auto generated process according to the Per Feature Flags struct being updated as a source of truth.
```

I still do not follow this statement 🙃

Reviewer 2: I think it means that the table of feature flags and their stability level should be generated automatically from the code. I think this is a good idea - a slightly different approach could be to use a json or yaml file to track the level of stability of the different features. The yaml could be used to generate docs as well as to select tests.

Author: Thanks for the question @pritidesai and clarification @afrittoli . Exactly, I wanted to say that this might be prone to error for `config-feature-flags` or just a Markdown files if people are editing it manually. Either a hack script or another json file could be very helpful in this case.

I also updated the wording for this future work.
